File: cpp/strategies/dual_ma.cpp
#include "strategies/dual_ma.h"
#include "core/engine.h"

namespace qe {

DualMA::DualMA(const std::string& symbol, int fast_period, int slow_period)
    : symbol_name_(symbol), fast_period_(fast_period), slow_period_(slow_period) {}

void DualMA::on_init(Context& ctx) {
    sid_ = ctx.symbol(symbol_name_);
    fast_ = &ctx.indicator<SMA>(sid_, fast_period_);
    slow_ = &ctx.indicator<SMA>(sid_, slow_period_);
}

void DualMA::on_bar(Context& ctx, uint16_t symbol_id, const Bar& bar) {
    if (symbol_id != sid_) return;
    if (!fast_->ready() || !slow_->ready()) return;

    auto& pos = ctx.position(sid_);

    if (fast_->value() > slow_->value() && pos.quantity == 0.0)
        ctx.buy(sid_, 0.1);
    else if (fast_->value() < slow_->value() && pos.quantity > 0.0)
        ctx.sell(sid_, pos.quantity);
}

void DualMA::on_stop(Context& ctx) {
    auto& pos = ctx.position(sid_);
    if (pos.quantity > 0)
        ctx.sell(sid_, pos.quantity);
}

}  // namespace qe

File: python/strategies/dual_ma.py
"""双均线策略 — Python 版

演示 Python 策略通过 pybind11 调用 C++ 引擎。
策略逻辑与 C++ 版 DualMA 完全一致：
  - fast SMA 上穿 slow SMA → 买入
  - fast SMA 下穿 slow SMA → 平仓
"""
import qe


class DualMA(qe.Strategy):
    def __init__(self, symbol_name: str, fast_period: int = 5, slow_period: int = 20):
        super().__init__()
        self._symbol_name = symbol_name
        self._fast_period = fast_period
        self._slow_period = slow_period
        self._sid = 0
        self._fast = None
        self._slow = None

    def on_init(self, ctx: qe.Context):
        self._sid = ctx.symbol(self._symbol_name)
        self._fast = ctx.sma(self._sid, self._fast_period)
        self._slow = ctx.sma(self._sid, self._slow_period)

    def on_bar(self, ctx: qe.Context, symbol_id: int, bar: qe.Bar):
        if symbol_id != self._sid:
            return
        if not self._fast.ready() or not self._slow.ready():
            return

        pos = ctx.position(self._sid)
        if self._fast.value() > self._slow.value() and pos.quantity == 0.0:
            ctx.buy(self._sid, 0.1)
        elif self._fast.value() < self._slow.value() and pos.quantity > 0.0:
            ctx.sell(self._sid, pos.quantity)

    def on_stop(self, ctx: qe.Context):
        pos = ctx.position(self._sid)
        if pos.quantity > 0:
            ctx.sell(self._sid, pos.quantity)
